pkg/cloud/cloud.go

```diff
@@ -18,9 +18,54 @@ limitations under the License.
 
 package cloud
 
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type uintslice []uint16
+
+func (i *uintslice) Type() string {
+	return "uintslice"
+}
+
+func (i *uintslice) String() string {
+	return fmt.Sprintf("%d", *i)
+}
+
+func (i *uintslice) Set(value string) error {
+	values := strings.Split(value, ",")
+
+	for _, val := range values {
+		tmp, err := strconv.ParseUint(val, 0, 16)
+		if err != nil {
+			return errors.Wrap(err, "Conversion to uint failed")
+		}
+
+		if !i.Contains(uint16(tmp)) {
+			*i = append(*i, uint16(tmp))
+		}
+	}
+
+	return nil
+}
+
+func (i *uintslice) Contains(value uint16) bool {
+	for _, port := range *i {
+		if value == port {
+			return true
+		}
+	}
+
+	return false
+}
+
```

Comment:
This bothers me: `cloud.Ports` is part of our Cobra-independent API, but we’re allowing Cobra-related concerns to leak into it here. We should change this to a `[]uint16` if that’s what we need, or a `[]uint` to make things simpler (since we’re changing all the metrics-port-handling code in `pkg/cloud/prepare` anyway, the latter isn’t a problem in practice, although it does mean more leaking of Cobra concerns into `pkg/cloud` albeit at a smaller scale).

Reply:
I can do the following:
* move the new type declaration `uintslice` to another file 
* use `uintslice` from `cmd\subctl\` 
* use `[]uint16` in `cloud.Ports`

Reply:
Again, I don’t like having Cobra concerns leak into this API. I would much rather have a `[]uint` here and `uint16()` casts as appropriate elsewhere.

Reply:
The only potential problem would be losing precision by casting if a user specified a large port number.  But, looking in **cloud-prepare**, the underlying usage of the `PortSpec.Port` field is varied, some convert to `int32`, some `int` and some to a `string`. I think we should just change  `PortSpec.Port` to `uint` and convert appropriately in the cloud provider code. Then we won't need to define a new Cobra type here.

Reply:
Ports have to be < 65536 anyway; we could add a check for that in the Cobra validations. Then as you say, change the ports to `uint`s.

Reply:
Check out [this commit](https://github.com/submariner-io/subctl/pull/74/commits/cf515b4453157fd6f31e3c6eb58f20f4c1dbb6c9). I think the new Cobra type is reasonable  as it validates but I moved it to `cmd/subctl`. The `Ports.Metrics` field is []uint16.

Reply:
Fair enough, thanks!